neuralop/layers/embeddings.py

```diff
@@ -137,35 +162,117 @@ def forward(self, data, batched=True):
+        "NeRF: Representing Scenes as Neural Radiance Fields for View Synthesis".
+        ArXiv, https://arxiv.org/pdf/2003.08934. 
+    """
+    def __init__(self, 
+                 in_channels: int,
+                 num_frequencies: int=None, 
+                 embedding_type: str='nerf', 
+                 max_positions: int=10000):
         super().__init__()
-        self.num_channels = num_channels
+        self.in_channels = in_channels
+        self.num_frequencies = num_frequencies
+        
+        # verify embedding type
+        allowed_embeddings = ['nerf', 'transformer']
+        assert embedding_type in allowed_embeddings, \
+            f"Error: embedding_type expected one of {allowed_embeddings}, received {embedding_type}"
+        self.embedding_type = embedding_type
         self.max_positions = max_positions
-        self.endpoint = endpoint
+    
+    
+    @property
+    def out_channels(self):
+        """
+        out_channels: required property for linking/composing model layers 
+        """
+        return 2 * self.num_frequencies * self.in_channels
 
     def forward(self, x):
-        freqs = torch.arange(
-            start=0, end=self.num_channels // 2, dtype=torch.float32, device=x.device
-        )
-        freqs = freqs / (self.num_channels // 2 - (1 if self.endpoint else 0))
-        freqs = (1 / self.max_positions) ** freqs
-        x = x.ger(freqs.to(x.dtype))
-        x = torch.cat([x.cos(), x.sin()], dim=1)
-        return x
-
+        """
+        Parameters 
+        -----------
+        x: torch.Tensor, shape (n_in, out_channels) or (batch, n_in, out_channels)
+        """
+        assert x.ndim in [2,3], f"Error: expected inputs of shape (batch, n_in, channels)\
+            or (n_in, channels), got inputs with ndim=={x.ndim}, shape={x.shape}"
+        if x.ndim == 2:
+            batched = False
+            x = x.unsqueeze(0)
+        else:
+            batched = True
+        batch_size, n_in, _ = x.shape
+        
+        if self.embedding_type == 'nerf':
+            freqs = 2 ** torch.arange(0, self.num_frequencies, device=x.device) * torch.pi
+        
+        elif self.embedding_type == 'transformer':
+            freqs = torch.arange(0, self.num_frequencies, device=x.device) / self.in_channels
+            freqs = (1 / self.max_positions) ** freqs
+        
+        # outer product of wavenumbers and position coordinates
+        # shape b, n_in * channels, len(freqs)
+        freqs = torch.einsum('bij, k -> bijk', x, freqs)
+
+        # shape len(x), 2, len(freqs)
+        freqs = torch.stack((freqs.sin(),freqs.cos()), dim=-1)
+
+        # transpose the inner per-entry matrix and ravel to interleave sin and cos
+        # for nerf, end result is (sin(2^0 * pi * p_0), 
+        #   cos(2^0 * pi * p_0), ...cos(2^k * pi * p_N))
+        # for transformer, end result is (sin(p_0/max_positions^{2k/N}), 
+        #   cos(p_0/max_positions^{2k/N}), ...cos(p_N/max_positions^{2k/N}))
```

Comment:
These annotations seem unclear. Maybe it is clearer to write something like:
nerf: (sin(2^0 * pi * p_0), ..., sin(2^num_frequencies * pi * p_{n_in-1})
        cos(2^0 * pi * p_0), ..., cos(2^num_frequencies * pi * p_{n_in-1}))
transformer: (sin(p_0/max_positions^{2*0/n_in}),  ..., sin(p_N/max_positions^{2*num_frequencies/n_in})
                      cos(p_0/max_positions^{2*0/n_in}), ..., cos(p_N/max_positions^{2*num_frequencies/n_in}))

Reply:
I think I'm just going to remove this line and clean up the refs at the top.